{"pr_number": 9700, "pr_title": "KAFKA-10813: InvalidProducerEpoch should be caught and throw as TaskMigrated", "pr_author": "abbccdda", "pr_createdAt": "2020-12-05T02:20:15Z", "pr_url": "https://github.com/apache/kafka/pull/9700", "timeline": [{"oid": "4cd4ea7c5e7dc784a64c1c688f1a5060a9487c11", "url": "https://github.com/apache/kafka/commit/4cd4ea7c5e7dc784a64c1c688f1a5060a9487c11", "message": "throw corresponding invalid producer epoch", "committedDate": "2020-12-06T06:29:08Z", "type": "commit"}, {"oid": "4cd4ea7c5e7dc784a64c1c688f1a5060a9487c11", "url": "https://github.com/apache/kafka/commit/4cd4ea7c5e7dc784a64c1c688f1a5060a9487c11", "message": "throw corresponding invalid producer epoch", "committedDate": "2020-12-06T06:29:08Z", "type": "forcePushed"}, {"id": "MDI0OlB1bGxSZXF1ZXN0UmV2aWV3Q29tbWVudDUzNzkxNTg5Nw==", "url": "https://github.com/apache/kafka/pull/9700#discussion_r537915897", "body": "```suggestion\r\n            } catch (final ProducerFencedException || InvalidProducerEpochException error) {\r\n```", "bodyText": "Suggested change\n        \n          \n    \n\n        \n      \n    \n    \n      \n          \n            \n                        } catch (final ProducerFencedException | InvalidProducerEpochException error) {\n          \n          \n            \n                        } catch (final ProducerFencedException || InvalidProducerEpochException error) {", "bodyHTML": "  <div class=\"my-2 border rounded-1 js-suggested-changes-blob diff-view js-check-bidi\" id=\"\">\n    <div class=\"f6 p-2 lh-condensed border-bottom d-flex\">\n      <div class=\"flex-auto flex-items-center color-fg-muted\">\n        Suggested change\n        <span class=\"tooltipped tooltipped-multiline tooltipped-s\" aria-label=\"This code change can be committed by users with write permissions.\">\n          <svg aria-hidden=\"true\" height=\"16\" viewBox=\"0 0 16 16\" version=\"1.1\" width=\"16\" data-view-component=\"true\" class=\"octicon octicon-info hide-sm\">\n    <path fill-rule=\"evenodd\" d=\"M8 1.5a6.5 6.5 0 100 13 6.5 6.5 0 000-13zM0 8a8 8 0 1116 0A8 8 0 010 8zm6.5-.25A.75.75 0 017.25 7h1a.75.75 0 01.75.75v2.75h.25a.75.75 0 010 1.5h-2a.75.75 0 010-1.5h.25v-2h-.25a.75.75 0 01-.75-.75zM8 6a1 1 0 100-2 1 1 0 000 2z\"></path>\n</svg>\n        </span>\n      </div>\n    </div>\n    <div itemprop=\"text\" class=\"blob-wrapper data file\" style=\"margin: 0; border: none; overflow-y: visible; overflow-x: auto;\">\n      <table class=\"d-table tab-size mb-0 width-full\" data-paste-markdown-skip=\"\">\n          <tbody><tr class=\"border-0\">\n            <td class=\"blob-num blob-num-deletion text-right border-0 px-2 py-1 lh-default\" data-line-number=\"192\"></td>\n            <td class=\"border-0 px-2 py-1 blob-code-inner blob-code-deletion js-blob-code-deletion blob-code-marker-deletion\">            } <span class=\"pl-k\">catch</span> (<span class=\"pl-k\">final</span> <span class=\"pl-smi\">ProducerFencedException</span> <span class=\"pl-k\">|</span> <span class=\"pl-smi\">InvalidProducerEpochException</span> error) {</td>\n          </tr>\n          <tr class=\"border-0\">\n            <td class=\"blob-num blob-num-addition text-right border-0 px-2 py-1 lh-default\" data-line-number=\"192\"></td>\n            <td class=\"border-0 px-2 py-1 blob-code-inner blob-code-addition js-blob-code-addition blob-code-marker-addition\">            } <span class=\"pl-k\">catch</span> (<span class=\"pl-k\">final</span> <span class=\"pl-smi\">ProducerFencedException</span> <span class=\"pl-k\">|<span class=\"x x-first x-last\">|</span></span> <span class=\"pl-smi\">InvalidProducerEpochException</span> error) {</td>\n          </tr>\n      </tbody></table>\n    </div>\n    <div class=\"js-apply-changes\"></div>\n  </div>\n", "author": "ableegoldman", "createdAt": "2020-12-07T23:33:16Z", "path": "streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsProducer.java", "diffHunk": "@@ -184,12 +184,12 @@ public void resetProducer() {\n         transactionInitialized = false;\n     }\n \n-    private void maybeBeginTransaction() throws ProducerFencedException {\n+    private void maybeBeginTransaction() {\n         if (eosEnabled() && !transactionInFlight) {\n             try {\n                 producer.beginTransaction();\n                 transactionInFlight = true;\n-            } catch (final ProducerFencedException error) {\n+            } catch (final ProducerFencedException | InvalidProducerEpochException error) {", "originalCommit": "4cd4ea7c5e7dc784a64c1c688f1a5060a9487c11", "replyToReviewId": null, "replies": [{"id": "MDI0OlB1bGxSZXF1ZXN0UmV2aWV3Q29tbWVudDUzODc1NTU3MQ==", "url": "https://github.com/apache/kafka/pull/9700#discussion_r538755571", "bodyText": "This won't work I suppose?", "author": "abbccdda", "createdAt": "2020-12-08T19:43:16Z", "replyToReviewId": "MDI0OlB1bGxSZXF1ZXN0UmV2aWV3Q29tbWVudDUzNzkxNTg5Nw=="}, {"id": "MDI0OlB1bGxSZXF1ZXN0UmV2aWV3Q29tbWVudDUzODg5OTc0Mw==", "url": "https://github.com/apache/kafka/pull/9700#discussion_r538899743", "bodyText": "Oh lol sorry I was just comparing it with the other block. Obviously nvm", "author": "ableegoldman", "createdAt": "2020-12-08T23:57:23Z", "replyToReviewId": "MDI0OlB1bGxSZXF1ZXN0UmV2aWV3Q29tbWVudDUzNzkxNTg5Nw=="}], "type": "inlineReview"}, {"id": "MDI0OlB1bGxSZXF1ZXN0UmV2aWV3Q29tbWVudDUzNzkxNzk1Mw==", "url": "https://github.com/apache/kafka/pull/9700#discussion_r537917953", "body": "Why is that here, we catch these three exceptions -- ProducerFenced, InvalidProducerEpoch, and OutOfOrderSequence -- and wrap them as TaskMigratedException, while in `StreamsProducer#send`, we catch  ProducerFenced, InvalidProducerEpoch, and UnknownProducerId exceptions and wrap those as TaskMigrated?", "bodyText": "Why is that here, we catch these three exceptions -- ProducerFenced, InvalidProducerEpoch, and OutOfOrderSequence -- and wrap them as TaskMigratedException, while in StreamsProducer#send, we catch  ProducerFenced, InvalidProducerEpoch, and UnknownProducerId exceptions and wrap those as TaskMigrated?", "bodyHTML": "<p dir=\"auto\">Why is that here, we catch these three exceptions -- ProducerFenced, InvalidProducerEpoch, and OutOfOrderSequence -- and wrap them as TaskMigratedException, while in <code>StreamsProducer#send</code>, we catch  ProducerFenced, InvalidProducerEpoch, and UnknownProducerId exceptions and wrap those as TaskMigrated?</p>", "author": "ableegoldman", "createdAt": "2020-12-07T23:38:17Z", "path": "streams/src/main/java/org/apache/kafka/streams/processor/internals/RecordCollectorImpl.java", "diffHunk": "@@ -199,7 +200,9 @@ private void recordSendError(final String topic, final Exception exception, fina\n         if (isFatalException(exception)) {\n             errorMessage += \"\\nWritten offsets would not be recorded and no more records would be sent since this is a fatal error.\";\n             sendException.set(new StreamsException(errorMessage, exception));\n-        } else if (exception instanceof ProducerFencedException || exception instanceof OutOfOrderSequenceException) {\n+        } else if (exception instanceof ProducerFencedException ||\n+                exception instanceof InvalidProducerEpochException ||\n+                exception instanceof OutOfOrderSequenceException) {", "originalCommit": "4cd4ea7c5e7dc784a64c1c688f1a5060a9487c11", "replyToReviewId": null, "replies": [{"id": "MDI0OlB1bGxSZXF1ZXN0UmV2aWV3Q29tbWVudDUzNzkyNDI0Ng==", "url": "https://github.com/apache/kafka/pull/9700#discussion_r537924246", "bodyText": "Just to clarify, I think my question/confusion is twofold:\n\nWhy is it OutOfOrderSequence in one place and UnknownProducerId in another?\nWhat is the difference between these two code paths? Is it really possible for example for the ProducerFencedException to sometimes be thrown directly from Producer#send, and sometimes be passed along through the callback ( as in streamsProducer.send(serializedRecord, (metadata, exception) )", "author": "ableegoldman", "createdAt": "2020-12-07T23:53:07Z", "replyToReviewId": "MDI0OlB1bGxSZXF1ZXN0UmV2aWV3Q29tbWVudDUzNzkxNzk1Mw=="}, {"id": "MDI0OlB1bGxSZXF1ZXN0UmV2aWV3Q29tbWVudDUzODY2NzY3Ng==", "url": "https://github.com/apache/kafka/pull/9700#discussion_r538667676", "bodyText": "Wasn't over thinking here since this is a blocker-fix, so we don't want to trigger any regression here by changing existing exception catching logic. Would be good to do this as a follow-up I guess.", "author": "abbccdda", "createdAt": "2020-12-08T17:59:12Z", "replyToReviewId": "MDI0OlB1bGxSZXF1ZXN0UmV2aWV3Q29tbWVudDUzNzkxNzk1Mw=="}, {"id": "MDI0OlB1bGxSZXF1ZXN0UmV2aWV3Q29tbWVudDUzODkwMDIwNA==", "url": "https://github.com/apache/kafka/pull/9700#discussion_r538900204", "bodyText": "Ok SG, agree we should keep things simple since this is a last minute blocker", "author": "ableegoldman", "createdAt": "2020-12-08T23:58:30Z", "replyToReviewId": "MDI0OlB1bGxSZXF1ZXN0UmV2aWV3Q29tbWVudDUzNzkxNzk1Mw=="}, {"id": "MDI0OlB1bGxSZXF1ZXN0UmV2aWV3Q29tbWVudDUzODk0MDU0NA==", "url": "https://github.com/apache/kafka/pull/9700#discussion_r538940544", "bodyText": "I thought about this a bit, and I think both UnknownProducerId and OutOfOrderSequence could be possibly thrown from the caller or directly (though in the later case they would be wrapped as KafkaException).\nI created two tickets, one for producer and one for streams to improve the general picture moving forward.\nhttps://issues.apache.org/jira/browse/KAFKA-10829\nhttps://issues.apache.org/jira/browse/KAFKA-10830", "author": "guozhangwang", "createdAt": "2020-12-09T01:43:14Z", "replyToReviewId": "MDI0OlB1bGxSZXF1ZXN0UmV2aWV3Q29tbWVudDUzNzkxNzk1Mw=="}], "type": "inlineReview"}]}